File: Csharp/RT_CS/RT_CS/shapes/Sphere.cs
using RT_CS.Cameras;
using RT_CS.utility;
using System;
using System.Collections.Generic;
using System.Text;

namespace RT_CS.shapes
{

    public class Sphere : Shape
    {
        private Vector3 centre;
        private float radius;

        public override Vector3 GetNormal(Vector3 point)
        {
            return (point - centre).Normalize();
        }

        public Sphere(int id,int mid, Vector3 c, float r) : base(id, mid)
        {
            centre = c;
            radius = r;
        }


        public override void Intersect(Ray r, IntersectionRecord record)
        {
            if (record.ExceptId == sid)
            {
                return;
            }
            Vector3 oc = r.posit - centre;
            float uOc = Vector3.Dot(r.direct, oc);
            float root = uOc * uOc - (oc.MagnitudeSq() - radius * radius);
            float d = -uOc;
            float dist = 0;
            if (Math.Abs(root) < float.Epsilon )
            {
                dist = d;
            }
            else if (root > float.Epsilon)
            {
                root = (float)Math.Sqrt(root);
                dist = Math.Min(d + root, d - root);
            }
            else
            {//< 0 means no intersection
                return;
            }

           
            if( dist > 0 && dist < record.Distance)
            {
                record.Distance =dist;
                record.Hit = true;
                record.Point = r.posit + dist * r.direct;
                record.ShapeId = sid;
            }
            
        }

        public override Vector2 GetUV(Vector3 pt)
        {
            Vector3 normal = GetNormal(pt);
            float u = MathF.Atan2(normal.x, normal.z) / (2 * MathF.PI) + 0.5F;
            float v = normal.y * 0.5F + 0.5F;
            return new Vector2() {  x = u, y = v };
        }
    }

}

File: cpp/RT_CPP/RT_CPP/sphere.cpp
#include "sphere.h"
#include "ray.h"
#include "operators.h"
#include <math.h>
# define M_PI           3.14159265358979323846 
#include <cfloat>
using namespace shapes;

sphere ::sphere (int id,int mid, vector3 c, float r) : shape(id,mid),
center(c), radius(r)
{

}

void sphere:: intersect(ray r, intersection_record& rec)  const
{
	if (id == rec.except_id)
	{
		return;
	}
	vector3 sub_1 = r.position - center;
	float sub_2 = vector3::dot(sub_1, r.direction);
	float grad = (sub_2 * sub_2) - (sub_1.magnitude_sq() - (radius * radius));
	float d = -sub_2;
	float dist = 0;
	if (fabsf(grad) < FLT_EPSILON)
	{
		dist = d;
	}

	else if (grad > FLT_EPSILON)
	{
		grad = sqrtf(grad);
		dist = fminf(d + grad, d - grad);
	}
	else
	{
		return;
	}
	
	if (dist > 0 && dist < rec.distance)
	{
		rec.hit = true;
		rec.distance = dist;
		rec.shape_id = id;
		rec.point = r.position + r.direction * dist;
	}
}

vector3 sphere::get_normal(vector3 point) const
{
	return (point - center).normalize();
}

vector2 sphere::get_uv(vector3 point) const
{
	vector3 normal = get_normal(point);
	float u = (atan2f(normal.x, normal.z) / (2 * M_PI)) + 0.5;
	float v = 0.5 - (asinf(normal.y) / M_PI);
	return { u,v };
}